internal/provider/edgegw/security_group_datasource.go

```diff
@@ -0,0 +1,118 @@
+// Package edgegw provides a Terraform datasource.
+package edgegw
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/hashicorp/terraform-plugin-framework/diag"
+
+	"github.com/hashicorp/terraform-plugin-framework/datasource"
+
+	"github.com/orange-cloudavenue/terraform-provider-cloudavenue/internal/client"
+	"github.com/orange-cloudavenue/terraform-provider-cloudavenue/internal/provider/common/edgegw"
+	"github.com/orange-cloudavenue/terraform-provider-cloudavenue/internal/provider/common/org"
+)
+
+var (
+	_ datasource.DataSource              = &securityGroupDataSource{}
+	_ datasource.DataSourceWithConfigure = &securityGroupDataSource{}
+)
+
+func NewSecurityGroupDataSource() datasource.DataSource {
+	return &securityGroupDataSource{}
+}
+
+type securityGroupDataSource struct {
+	client *client.CloudAvenue
+	org    org.Org
+	edgegw edgegw.EdgeGateway
+}
+
+// Init Initializes the data source.
+func (d *securityGroupDataSource) Init(ctx context.Context, dm *securityGroupModel) (diags diag.Diagnostics) {
+	var err error
+
+	d.org, diags = org.Init(d.client)
+	if diags.HasError() {
+		return
+	}
+
+	d.edgegw, err = d.org.GetEdgeGateway(edgegw.BaseEdgeGW{
+		ID:   dm.EdgeGatewayID,
+		Name: dm.EdgeGatewayName,
+	})
+	if err != nil {
+		diags.AddError("Error retrieving Edge Gateway", err.Error())
+		return
+	}
+
+	return
+}
+
+func (d *securityGroupDataSource) Metadata(ctx context.Context, req datasource.MetadataRequest, resp *datasource.MetadataResponse) {
+	resp.TypeName = req.ProviderTypeName + "_" + categoryName + "_security_group"
+}
+
+func (d *securityGroupDataSource) Schema(ctx context.Context, req datasource.SchemaRequest, resp *datasource.SchemaResponse) {
+	resp.Schema = securityGroupSchema(ctx).GetDataSource(ctx)
+}
+
+func (d *securityGroupDataSource) Configure(ctx context.Context, req datasource.ConfigureRequest, resp *datasource.ConfigureResponse) {
+	// Prevent panic if the provider has not been configured.
+	if req.ProviderData == nil {
+		return
+	}
+
+	client, ok := req.ProviderData.(*client.CloudAvenue)
+	if !ok {
+		resp.Diagnostics.AddError(
+			"Unexpected Data Source Configure Type",
+			fmt.Sprintf("Expected *client.CloudAvenue, got: %T. Please report this issue to the provider developers.", req.ProviderData),
+		)
+		return
+	}
+	d.client = client
+}
+
+func (d *securityGroupDataSource) Read(ctx context.Context, req datasource.ReadRequest, resp *datasource.ReadResponse) {
+	config := &securityGroupModel{}
+
+	// Read Terraform configuration data into the model
+	resp.Diagnostics.Append(req.Config.Get(ctx, config)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	// Init the resource
+	resp.Diagnostics.Append(d.Init(ctx, config)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	/*
+		Implement the data source read logic here.
+	*/
+
+	s := &securityGroupResource{
+		client: d.client,
+		org:    d.org,
+		edgegw: d.edgegw,
+	}
```

Comment:
why not use d *securityGroupDataSource ?

Reply:
Because all func getSecurityGroup/read is already available in Resource. 
This technic is already used in firewall datasource https://github.com/orange-cloudavenue/terraform-provider-cloudavenue/blob/main/internal/provider/edgegw/firewall_datasource.go#L96

Reply:
You set resource struct with datasource information, i think it's better to rename the struct with a generic name without the word resource or precise in commentary your trick.